### DISCUSSION TITLE
Public API - Issues With Multiple Filters For Key

### DISCUSSION THREAD

**Discussion by brocococonut** [Budibase/budibase#6615] [Help]
## Short version:
As the title says, I'm currently having issues with using multiple filters for a single key when querying the public API. Additionally, using Dynamic Filters on the frontend and selecting multiple Equals or NotEquals filters for any given key results in only one being used. This raises concern over whether regular data-provider filters behave as expected in general. Additionally, these limitations (if intentional) aren't specified anywhere in the code or in the docs (or API reference) from what I can tell.

## Expanded details
It makes sens that a query object would look like the following:
```json
{
  "query": {
    "notEqual": {
      "newKey": "New Value",
      "newKey-2": "New Values"
    }
  }
}
```

This however raises some issues when it comes to querying for multiple NotEquals (or other filter type) conditions on data. If you create a dynamic filter object on a page and do something like the following:
![image](https://user-images.githubusercontent.com/6368283/177983498-00dc5e47-8c4f-40a6-b267-5801d2838759.png)
and monitor network requests, this is the filter that gets applied:
```json
{
  "query": {
    "notEqual": {
      "Outcome": "Cancelled At Door"
    }
  }
}
```

As we can see, one of the conditions was just stripped away as it didn't fit the `{[key: string]: string}` format. This leads me to believe the same is going to happen anywhere else in the app that uses filters. This is either a design oversight or something that was just ignored when development of the public (and likely internal) API was performed. Something simple like the following would almost definitely result in the same issue from what I can tell:
![image](https://user-images.githubusercontent.com/6368283/177984186-9e0ff3a6-3917-425d-8bf7-498793874ed6.png)

A suggested alternative to the existing query format could be something like the following:
```jsonc
{
  "query": [
    {
      "notEqual": {
        "Outcome": "Cancelled At Door"
      }
    },
    {
      "notEqual": {
        "Outcome": "Other"
      }
    }
  ]
}

// Or
{
  "query": {
    "notEqual": [
      {
        "Outcome": "Cancelled At Door"
      },
      {
        "Outcome": "Other"
      }
    ]
  }
}

// Or
{
  "query": {
    "notEqual": {
      "Outcome": [
        "Cancelled At Door",
        "Other"
      ]
    }
  }
}
```
All of which have various drawbacks/issues in readability or verboseness, but I feel like any one of them would be less restrictive on things.

The only alternative I can see at the moment is doing a bigger broad request where there aren't any overlapping queries, and then filtering them after the fact.

Any thoughts?

**Comment by brocococonut** [comment]
An alternative solution to the above would be an `in` and `notIn` filter option so you can do essentially the same thing without rewriting how the `equals` and `notEquals` code works - which would in turn negate any changes to the docs for those filters.

**Reply by mike12345567** [reply]
Hi @brocococonut - this is a known issue, https://github.com/Budibase/budibase/issues/2585 - it is something we hope to fix at some point, at the moment `oneOf` filters can be applied for SQL tables - the next version of Budibase will also support `oneOf` on internal tables.
